Editors/vscode/package.json

```diff
@@ -68,6 +68,19 @@
                         "verbose"
                     ],
                     "description": "Traces the communication between VS Code and the SourceKit-LSP language server."
+                },
+                "sourcekit-lsp.indexVisibility.targets": {
+                    "type": "array",
+                    "default": [],
+                    "items": {
+                        "type": "string"
+                    },
+                    "description": "Restrict visibility to targets. Use when SourceKit-LSP language server is launched with `-explicit-index-loading` option"
```

Comment:
The indexstore database initialization happens earlier than workspace setting notification, so automatic explicit mode setting might not apply here.

Reply:
Is your concern that the user might change the configuration from [] to [<target>] and we cannot enable the explicit mode at that point?  Isn't that already a problem with the current approach, since they cannot add initialization options without restarting the service?

It's also a bit unfortunate that once you turn this on there is no way to disable it without restarting (e.g. changing the target list to `nil` or `[]` will make it so nothing is visible rather than everything).  @akyrtzi is there a fundamental reason that we cannot dynamically enable/disable `useExplicitOutputUnits`?

Reply:
Right now we receive the configuration through configurationDidChange notification sent by the ide. This happens later than workspace (and index database) initialization, and since we can't flip `useExplicitOutputUnits` after initialization, this is not doable.

If we want to, we could block the indexdb initialization by sending  a `workspace/configuration` request right after initialization response, but that comes with a bit of start up overhead.

Reply:
>  @akyrtzi is there a fundamental reason that we cannot dynamically enable/disable useExplicitOutputUnits?

It could be implemented but switching between the modes would be more complicated and less clean. The difference is not just a 'visibility' check, "explicit" mode allows to completely avoid reading any units that are not included, so it works a bit differently from that respect.

Reply:
What would you recommend? If we expose this in the vscode extension, I think users would expect to be able to change it, including going from filtered to unfiltered, or vice versa, and have that reflected.  It seems like we have the following options

1. Expose this only as an initialization-time option, and require restarting the service to see a change, the approach that this patch takes.  The downside is it's not very user-friendly to enable or disable.
2. Always have filtering on in the index, but default to looking up all targets and making them visible by querying the build system.
3. Implement dynamically enabling/disabling filtering inside the index
4. Like (3), but do it by closing/reopening the DB (note that reopening the DB after turning off filtering will not remove things from the DB unless we throw it out entirely).

If we want to do (2), we need to implement the target queries that we're using in the swiftpm build system and the compilation database build system. For a compilation database I think you just get a single target that has all the object files.

Reply:
I think the current approach (1) set the correct expectation for user in terms of explicit mode can only be changed during initialization. Are you concerned that user expect global index visibility when there's no targets set?

2 and 4 might introduce some performance overhead, depending on the build system and DB on/off overhead
3 is definitely preferred though as @akyrtzi suggested that might requires some effort.

Reply:
This doesn't seem to me like something that users would need to frequently change while working on a workspace, you would normally set it once and use it always (or more likely the client/IDE chooses appropriately for a workspace). I think #1 is acceptable IMO.